### ISSUE TITLE
ZonedDateTime to XmlGregorianCalendar losses time zone information

### ISSUE THREAD

**Issue by cloose** [mapstruct/mapstruct#1523]
Mapstruct: **1.2.0.Final**
Java: **JDK 1.8.0-45**

# Issue

We unfortunately have an issue with the mapping from `ZonedDateTime` to `XMLGregorianCalendar` and we believe this problem comes from the generated MapStruct code.

During the mapping the time zone information that was set on the ZonedDateTime is lost and the resulting XmlGregorianCalendar has the default system timezone.

Entity PickupDate:
~~~java
public class PickupDate {

    private final LocalDate value;
    private final ZonedDateTime validUntil;

    public PickupDate(LocalDate value, ZonedDateTime validUntil) {
        Assert.notNull(value, "value must not be null");
        Assert.notNull(validUntil, "validUntil must not be null");

        this.value = value;
        this.validUntil = validUntil;
    }

    public LocalDate getValue() {
        return value;
    }

    public ZonedDateTime getValidUntil() {
        return validUntil;
    }
}
~~~

Generated XML Type PickupDateType:
~~~java
@XmlAccessorType(XmlAccessType.FIELD)
@XmlType(name = "PickupDateType", propOrder = {
    "value",
    "validUntil"
})
public class PickupDateType {

    @XmlElement(name = "Value", required = true)
    @XmlSchemaType(name = "date")
    protected XMLGregorianCalendar value;
    @XmlElement(name = "ValidUntil", required = true)
    @XmlSchemaType(name = "dateTime")
    protected XMLGregorianCalendar validUntil;

    public XMLGregorianCalendar getValue() {
        return value;
    }

    public void setValue(XMLGregorianCalendar value) {
        this.value = value;
    }

    public XMLGregorianCalendar getValidUntil() {
        return validUntil;
    }

    public void setValidUntil(XMLGregorianCalendar value) {
        this.validUntil = value;
    }
}
~~~

Mapper interface:

~~~java
@Mapper(disableSubMappingMethodsGeneration = true)
public interface PickupDateTypeMapper {

    PickupDateType toPickupDateType(PickupDate pickupDate);

    List<PickupDateType> toPickupDateTypeList(List<PickupDate> pickupDates);

}
~~~

In the mapper implementation the following line is generated to set the XMLGregorianCalendar:
~~~java
pickupDateType.setValidUntil( calendarToXmlGregorianCalendar( zonedDateTimeToCalendar( pickupDate.getValidUntil() ) ) );
~~~

The method `zonedDateTimeToCalendar` returns a GregorianCalendar with the correct time zone information:
~~~java
    private Calendar zonedDateTimeToCalendar(ZonedDateTime dateTime) {
        if ( dateTime == null ) {
            return null;
        }

        Calendar instance = Calendar.getInstance( TimeZone.getTimeZone( dateTime.getZone() ) );
        instance.setTimeInMillis( dateTime.toInstant().toEpochMilli() );
        return instance;
    }
~~~

But then the `calendarToXmlGregorianCalendar` method is used and it creates a `GregorianCalendar` from the passed `Calendar` without preserving the time zone information.
~~~java
    private XMLGregorianCalendar calendarToXmlGregorianCalendar( Calendar cal ) {
        if ( cal == null ) {
            return null;
        }

        try {
            GregorianCalendar gcal = new GregorianCalendar();
            gcal.setTimeInMillis( cal.getTimeInMillis() );
            return DatatypeFactory.newInstance().newXMLGregorianCalendar( gcal );
        }
        catch ( DatatypeConfigurationException ex ) {
            throw new RuntimeException( ex );
        }
    }
~~~

# Suggested Solution

A direct mapping between `ZonedDateTime` and `XMLGregorianCalendar` should be provided instead of the current two mapping methods:
~~~java
    private XMLGregorianCalendar zonedDateTimeToXmlGregorianCalendar(ZonedDateTime dateTime) {
        if ( dateTime == null ) {
            return null;
        }

        try {
            return DatatypeFactory.newInstance().newXMLGregorianCalendar(GregorianCalendar.from(zonedDateTime));
        } catch (DatatypeConfigurationException ex) {
            throw new RuntimeException( ex );
        }
    }
~~~

Thank you for your great software!
Christian Loose

**Comment by chris922** [comment]
I just added a PR #1525 that should solve this issue.

**Comment by sjaakd** [comment]
done.. fixed by merged PR #1525
